x/auth/ante/fee.go

```diff
@@ -0,0 +1,99 @@
+package ante
+
+import (
+	"fmt"
+	sdkante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/e-money/em-ledger/x/buyback"
+)
+
+// This is a fork of the DeductFeeDecorator from the SDK version v0.42.4
+// It deposits the fees differently based on what denomination it is paid with.
+// If the fee is paid in NGM, it is sent to the tradition fee pool and distributed as rewards
+// If the fee is paid with a stablecoin balance, it is sent to the buyback module
+// https://github.com/e-money/em-ledger/issues/41
+type DeductFeeDecorator struct {
+	ak            sdkante.AccountKeeper
+	bankKeeper    types.BankKeeper
+	stakingKeeper StakingKeeper
+}
+
+func NewDeductFeeDecorator(ak sdkante.AccountKeeper, bk types.BankKeeper, sk StakingKeeper) DeductFeeDecorator {
+	return DeductFeeDecorator{
+		ak:            ak,
+		bankKeeper:    bk,
+		stakingKeeper: sk,
+	}
+}
+
+func (dfd DeductFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	if addr := dfd.ak.GetModuleAddress(types.FeeCollectorName); addr == nil {
+		panic(fmt.Sprintf("%s module account has not been set", types.FeeCollectorName))
+	}
+
+	if addr := dfd.ak.GetModuleAddress(buyback.AccountName); addr == nil {
+		panic(fmt.Sprintf("%s module account has not been set", buyback.AccountName))
+	}
+
+	feePayer := feeTx.FeePayer()
+	feePayerAcc := dfd.ak.GetAccount(ctx, feePayer)
+
+	if feePayerAcc == nil {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnknownAddress, "fee payer address: %s does not exist", feePayer)
+	}
+
+	// deduct the fees
+	if !feeTx.GetFee().IsZero() {
+		err = deductFees(dfd.bankKeeper, dfd.stakingKeeper, ctx, feePayerAcc, feeTx.GetFee())
+		if err != nil {
+			return ctx, err
+		}
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+// deductFees deducts fees from the given account.
+func deductFees(bankKeeper types.BankKeeper, stakingKeeper StakingKeeper, ctx sdk.Context, acc types.AccountI, fees sdk.Coins) error {
+	if !fees.IsValid() {
+		return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFee, "invalid fee amount: %s", fees)
+	}
```

Comment:
What if the user does not specify fees? According to the SDK 
> they must indicate 2 of the 3 following parameters (the third one being implicit): fees, gas and gas-prices.

Reply:
Actually the --gas, --gas-prices are factored in feeTx.GetFees(), feeTx.GetGas().